### PR TITLE
Adding subject metadata and ruleset capabilities

### PR DESCRIPTION
**Pull request by StephanePaulus**
Adding support for subject metadata, rulesets and schema normalisation.

This is also supported in the maven plugin. 

What I did is a very rudimentary implementation, if you have any ideas. I am happy to have a look and modify the code. 

Referencing a file, is an easy way, else I am afraid that the Gradle extension will become to complicated. 

Also my Kotlin experience is limited.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -29,6 +30,7 @@ dependencies {
     implementation("io.confluent", "kafka-protobuf-provider")
     // Our custom avro version. See https://github.com/ImFlog/avro
     implementation("com.github.ImFlog.avro", "avro", avroVersion)
+    implementation("com.google.code.gson","gson",jsonVersion)
```
Comment (ImFlog) [inline build.gradle.kts:33]: `gson` is already available as a transitive dependency from kafka-schema-registry-parent.
Thus there is no need to define the version.

Also i term of formatting, add spaces after the `,`.

File: src/main/kotlin/com/github/imflog/schema/registry/Subject.kt [hunk 2]
```diff
@@ -9,6 +13,10 @@ data class Subject(
 ) {
     val references: MutableList<SchemaReference> = mutableListOf()
     val localReferences: MutableList<LocalReference> = mutableListOf()
+    var metadata: Metadata = Metadata(mutableMapOf(), mutableMapOf(), mutableSetOf())
+    var ruleSet: RuleSet = RuleSet(mutableListOf(), mutableListOf())
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/Subject.kt:17]: I think the default value should be null values.
Like this we would not register anything in the rules / metadata if nothing is provided.

File: src/test/kotlin/com/github/imflog/schema/registry/tasks/register/RegisterTaskActionTest.kt [hunk 3]
```diff
@@ -337,6 +337,7 @@ class RegisterTaskActionTest {
         val subjects = listOf(
             Subject("test", "${schemaPath}testSubject.avsc", "AVRO")
                 .addLocalReference("B", "${schemaPath}testType.avsc")
+                .setRuleSet("${schemaPath}testRuleSet.json")
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/tasks/register/RegisterTaskActionTest.kt:340]: I think we should add a dedicated test that includes testing on the ruleSet, metadata and normalization.

File: src/main/kotlin/com/github/imflog/schema/registry/tasks/register/RegisterTaskAction.kt [hunk 4]
```diff
@@ -27,9 +29,9 @@ class RegisterTaskAction(
     fun run(): Int {
         var errorCount = 0
         writeOutputFileHeader()
-        subjects.forEach { (subject, path, type, references: List<SchemaReference>, localReferences) ->
+        subjects.forEach { (subject, path, type, references: List<SchemaReference>, localReferences, metaData, ruleSet, normalize) ->
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/tasks/register/RegisterTaskAction.kt:32]: We start to have too many things de-structured. We probably should go back to use a variable and use accessors.

File: src/main/kotlin/com/github/imflog/schema/registry/Subject.kt [hunk 5]
```diff
@@ -25,7 +33,27 @@ data class Subject(
...
+    }
+
     // Used for data class destructuring
     operator fun component4() = references
     operator fun component5() = localReferences
-}
\ No newline at end of file
+    operator fun component6() = metadata
+    operator fun component7() = ruleSet
+    operator fun component8() = normalize
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/Subject.kt:58]: As explained later, I think we should remove the de-structuring operator and go back to using a variable.

**Review by ImFlog — CHANGES_REQUESTED** [review]
Thank you for the PR ! This seems to be a valid new feature that I didn't knew about because it's only available for Confluent customers.

Few things that comes to my mind in addition to the review for the current code:
* We should update the README to reflech the new usage
* Add a separate test for rules, metadata and normalization
* Add an integration test
* Update the example repository to show how it could work

Also, It seems possible to retrieve those data when downloading the schema, we should probably add a flag to save those data as well if the user wants to.
WDYT ?

**Comment by StephanePaulus** [comment]
I did the code changes you asked for and added some quick tests. I will look into adding an integration test, example and readme for the changes. 

About the downloading of the schema, this data will already be returned when also downloading the optional metadata. 

    download {
        // Optional
        metadata = new MetadataExtension(true, "path/to/metadata/")

        // extension of the output file depends on the the schema type
        subject('avroSubject', '/absolutPath/src/main/avro')
        subject('protoSubject', 'src/main/proto')
        subject('jsonSubject', 'src/main/json')

        // You can use a regex to download multiple schemas at once
        subjectPattern('avro.*', 'src/main/avro')
    }

**Comment by StephanePaulus** [comment]
After looking into writing integration-test, the ruleSet and metadata are not being downloading when using `confluentinc/cp-schema-registry`, I guess this only works for confluent-cloud.

**Comment by StephanePaulus** [comment]
Hey @ImFlog, do you think the changes I have made are sufficient?

File: examples/avro-kts/build.gradle.kts [hunk 6]
```diff
@@ -17,6 +17,10 @@ schemaRegistry {
                 .addReference("company", "company", -1)
         subject("location-address", "schemas/avro/location-address.avsc", "AVRO")
         subject("location-latlong", "schemas/avro/location-latlong.avsc", "AVRO")
+        subject("my-record","schemas/avro/my-record.avsc", "AVRO")
+            .setMetadata("examples/extra/metadata.json")
+            .setRuleSet("examples/extra/ruleSet.json")
+            .setNormalized(true)
```
Comment (ImFlog) [inline examples/avro-kts/build.gradle.kts:23]: Based on your comment with confluent cloud, does it work correctly when you run the example directory ?
If not we could comment those line with a short message like you did in the README ?

**Review by ImFlog — CHANGES_REQUESTED** [review]
Sorry for the late answer 🙇 
Nearly all good, few remarks and also can you add an Integration test too (just to verify gradle is happy without looking at the answer).
Thank you again !

File: examples/avro-kts/build.gradle.kts (hunk 6 above)
Comment (StephanePaulus) [reply examples/avro-kts/build.gradle.kts:23]: The normalize and metadata work download, ruleset will always return `null` when downloaded, but registering it is fine. I guess the schema registry just ignores it. If you want me to comment it out let me know.

File: build.gradle.kts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 
 group = "com.github.imflog"
-version = "1.13.1-SNAPSHOT"
+version = "1.13.4-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (StephanePaulus) [reply build.gradle.kts:4]: Sorry, I changed it again.

File: examples/avro-kts/build.gradle.kts (hunk 6 above)
Comment (ImFlog) [reply examples/avro-kts/build.gradle.kts]: If it doesn't fail then it's good to keep it I think 👍

**Comment by ImFlog** [comment]
LGTM !
There is just one test that is failing, when fixed we will be able to merge this 😄

**Review by ImFlog — APPROVED** [review]

**Pull request merged**